### PR TITLE
PCH Excerpt Suggestions: Add Persona and Tone settings

### LINKED ISSUE #2692: Excerpt Suggestions: Allow different personas/styles
## Is your feature request related to a problem?
In Excerpt Suggestions, we currently don't offer a way to customize the persona and style. We recently received internal feedback that this would be desirable, since the current persona/style isn't ideal for writings of certain types (e.g. documentation).

## Describe the solution you'd like
Provide controls to specify persona and style, as we currently do in our Title Suggestions feature.

### PR DESCRIPTION
**Pull request by acicovic**
## Description
This PR adds `Persona` and `Tone` settings to our `Excerpt Suggestions` feature, in a similar manner to what already exists in our `Title Suggestions` feature.

## Motivation and context
- Users want to be able to customize the persona and tone of generated excerpts.
- Closes #2692.

### CODE REVIEW

**Comment by acicovic** [comment]
Notes:
- This work is currently based on our [refactor/internal-api](https://github.com/Parsely/wp-parsely/tree/refactor/internal-api) branch in order to avoid duplicate work since it adds a settings endpoint.
- @vaurdan, trying different things, I still couldn't make `SettingsProvider` to work. I suspect it might have to do with this being in Gutenberg Settings, and maybe a higher order component is needed as you've done previously for Smart Linking. Nothing has worked for me so far, but I'm probably missing something. Could you take a look?
- This PR is currently incomplete due to the previous point.

**Comment by acicovic** [comment]
@vaurdan: regarding my previous comment, let's also keep in mind that there's a feature request to have [Excerpt Suggestions in the PCH Sidebar](https://github.com/Parsely/wp-parsely/issues/2552).

As this will probably affect how Excerpt Suggestions settings get saved, it would be good to factor this in during this work. If we're unsure of how this will play out, let's start implementing that request to avoid duplicate/refactoring work.

**Comment by vaurdan** [comment]
@acicovic I went ahead and added the SettingsProvider to the component. It was a bit trickier, because of how we're using a SlotFill to insert the Excerpt panel to the Document sidebar (Using [PluginDocumentSettingPanel](https://developer.wordpress.org/block-editor/reference-guides/slotfills/plugin-document-setting-panel/)). 

This required a bit of reworking on how the code is organized - instead of being a standalone module, I made it part of the `editor-sidebar` module. This will also help when we add the panel to the PCH sidebar, since that move would have to be done nonetheless. 

From your original code, instead of having a new settings namespace for the excerpt generator, I changed it so it's also part of the main SidebarSettings namespace. 

Let me know your thoughts, and if you have any questions!

**Comment by vaurdan** [comment]
Oh - I also renamed the component from `excerpt-generator` to `excerpt-suggestions` on a different commit 🙂

**Review by 69023030 — APPROVED** [review]

File: src/content-helper/editor-sidebar/excerpt-suggestions/class-excerpt-suggestions.php
Comment (acicovic) [inline src/content-helper/editor-sidebar/excerpt-suggestions/class-excerpt-suggestions.php]: I think this will break existing usages of the `wp_parsely_enable_content_helper_excerpt_generator` filter. A strategy here would be to deprecate the old filter name, but still support if for some time. We should do the same if we've changed other filter names.

File: src/content-helper/editor-sidebar/excerpt-suggestions/class-excerpt-suggestions.php
Comment (acicovic) [inline src/content-helper/editor-sidebar/excerpt-suggestions/class-excerpt-suggestions.php]: Has this `@since` been updated on purpose? Same for the one below.

File: src/content-helper/common/settings/types/sidebar-settings.d.ts
Comment (acicovic) [inline src/content-helper/common/settings/types/sidebar-settings.d.ts]:
```suggestion
 * Defines the settings structure for the PostExcerptSuggestions component.
```

**Review by acicovic — COMMENTED** [review]
Thank you for completing this work! I agree with renaming ExcerptGenerator to ExcerptSuggestions anywhere possible.

I've left a couple of comments, the only really important one being the one about the filter.

**Review by 69023030 — COMMENTED** [review]
T

File: src/content-helper/editor-sidebar/excerpt-suggestions/class-excerpt-suggestions.php
Comment (vaurdan) [reply src/content-helper/editor-sidebar/excerpt-suggestions/class-excerpt-suggestions.php]: Good call!

Comment (vaurdan) [reply src/content-helper/editor-sidebar/excerpt-suggestions/class-excerpt-suggestions.php]: I did addressed this on [b8e3978](https://github.com/Parsely/wp-parsely/pull/2776/commits/b8e3978b948c87578da32a098db3e4a6aacfe9bb). I ended up reverting back to the `excerpt_generator`, and it should send a deprecation notice (if WP_DEBUG is enabled), using `apply_filters_deprecated` and suggesting `wp_parsely_current_user_can_use_pch_feature`.

This required some tweaking since the `wp_parsely_enable_content_helper_excerpt_generator` was not even working properly, as now the excerpt generator does not have it's own JS and CSS files, so we can't simply not load it. So I hooked up that filter with the new permission feature of the plugin, so if `wp_parsely_enable_content_helper_excerpt_generator` is set and returning false, is as if the feature was disabled.

**Comment by vaurdan** [comment]
@acicovic I believe I addressed all your feedback! Let me know if you have any additional thoughts 🙂 

I believe since you're the owner of this PR, you can't approve it, but if we have 🟢 , let me know and I can approve it myself.

File: src/content-helper/editor-sidebar/class-editor-sidebar-feature.php
Comment (acicovic) [inline src/content-helper/editor-sidebar/class-editor-sidebar-feature.php:64]: Extra blank line.

```suggestion
```

File: src/content-helper/editor-sidebar/class-editor-sidebar-feature.php
Comment (acicovic) [inline src/content-helper/editor-sidebar/class-editor-sidebar-feature.php:78]: Extra blank line.

```suggestion
```

File: src/content-helper/editor-sidebar/class-editor-sidebar-feature.php
Comment (acicovic) [inline src/content-helper/editor-sidebar/class-editor-sidebar-feature.php]: Is this `@since` correct?

File: src/content-helper/editor-sidebar/class-editor-sidebar.php
Comment (acicovic) [inline src/content-helper/editor-sidebar/class-editor-sidebar.php]: Extra tab that seems unneed?

```suggestion
			$feature->run();
```

File: src/content-helper/editor-sidebar/excerpt-suggestions/class-excerpt-suggestions.php
Comment (acicovic) [inline src/content-helper/editor-sidebar/excerpt-suggestions/class-excerpt-suggestions.php:97]: I think it would be nice to have a blank line above the return.

```suggestion

		return $current_user_can_use_pch_feature;
```

File: src/content-helper/editor-sidebar/excerpt-suggestions/class-excerpt-suggestions.php
Comment (acicovic) [inline src/content-helper/editor-sidebar/excerpt-suggestions/class-excerpt-suggestions.php]: This function returns a bool flag, but it's name indicates that it disables the feature. Maybe we can find a name that's a better fit for it?

**Review by acicovic — COMMENTED** [review]
Thanks for addressing the filters situation!

I've left some minor comments.

File: src/content-helper/editor-sidebar/excerpt-suggestions/class-excerpt-suggestions.php
Comment (vaurdan) [reply src/content-helper/editor-sidebar/excerpt-suggestions/class-excerpt-suggestions.php]: Yeah, I struggled a little bit trying to find the most appropriate name for this callback function. I renamed it to `callback_is_feature_enabled`, since it hooks to `wp_parsely_current_user_can_use_pch_feature`. What do you think?

Comment (acicovic) [reply src/content-helper/editor-sidebar/excerpt-suggestions/class-excerpt-suggestions.php]: Yes, `callback_is_feature_enabled` makes sense to me!

Comment (acicovic) [reply src/content-helper/editor-sidebar/excerpt-suggestions/class-excerpt-suggestions.php]: Function summary says `Checks and disables the feature if it is disabled via the global or feature filters.`, which isn't true since this function doesn't disable anything.

**Comment by acicovic** [comment]
I've left a comment in our last discussion, but otherwise I think we're good to go. Let's not merge this though, until our API refactoring branch is merged first.

**Comment by acicovic** [comment]
Closing in favor of #2890.

**Pull request closed without merging**
